backend/regalloc/regalloc_split_state.ml

```diff
@@ -51,6 +51,157 @@ let log_renaming_info : t -> unit =
   dedent ();
   dedent ()
 
+(* Optimizes `destructions_at_end` and `definitions_at_beginning`, by moving
+   destructions and definitions outside of loops when possible. *)
+module ExtractSpillsAndReloadsFromLoops : sig
+  val optimize :
+    Cfg_with_infos.t ->
+    destructions_at_end:destructions_at_end ->
+    definitions_at_beginning:definitions_at_beginning ->
+    destructions_at_end * definitions_at_beginning
+end = struct
+  type loop_sets =
+    { destroyed : Reg.Set.t;
+      occurring : Reg.Set.t
+    }
+
+  let compute_loop_sets cfg loop destructions_at_end =
+    Label.Set.fold
+      (fun label { destroyed; occurring } ->
+        let destroyed_here =
+          match Label.Map.find_opt label destructions_at_end with
+          | None -> Reg.Set.empty
+          | Some (_kind, regs) -> regs
+        in
+        let destroyed = Reg.Set.union destroyed destroyed_here in
+        let block = Cfg.get_block_exn cfg label in
+        let[@inline] update_occuring :
+            type a. Reg.Set.t -> a Cfg.instruction -> Reg.Set.t =
+         fun occurring instr ->
```

Comment:
This doesn't need to add regs destroyed by the instruction to `destroyed` because those are always hardware regs, right? Could use a comment.